torch/distributed/_shard/sharded_tensor/api.py

```diff
@@ -62,7 +61,53 @@ def _register_remote_shards(sharded_tensor_id: int, rrefs: List[rpc.RRef[Shard]]
         else:
             sharded_tensor._register_remote_shards(rrefs, rpc_rank)
 
-class ShardedTensor(object):
+
+class ShardedTensorInterface(torch.Tensor):
+    @staticmethod
+    def __new__(cls,
+                sharding_spec: shard_spec.ShardingSpec,
+                *size,
+                **kwargs):
+        # Use __new__ for logging purposes.
+        torch._C._log_api_usage_once("torch.distributed._shard.sharded_tensor")
+        sizes = _flatten_tensor_size(size)
+        dtype = kwargs['dtype']
+        layout = kwargs['layout']
+        pin_memory = kwargs['pin_memory']
+        requires_grad = kwargs['requires_grad']
+        r = torch.Tensor._make_wrapper_subclass(
+            cls,
+            sizes,
+            dtype=dtype,
+            layout=layout,
+            pin_memory=pin_memory,
+            requires_grad=requires_grad
+        )
+        return r
+
+    # We define this function for two reasons:
+    #  - So that this subclass is recognised as a python subclass by the backend
+    #  - So that the user gets friendly errors if they use only torch_function but
+    #    their subclass is used on the c++ side (by autograd for example)
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        raise RuntimeError(f"A {cls.__name__} object is being used from c++ while calling {func.__module__}.{func.__name__} "
+                           "but the there is no custom __torch_dispatch__ implementation for it.")
+
+    def sharding_spec(self) -> shard_spec.ShardingSpec:
+        raise NotImplementedError()
+
+    def local_tensor(self) -> torch.Tensor:
+        raise NotImplementedError()
```

Comment:
Maybe we remove `local_tensor` from the interface since it is mostly a specialization of `local_shards`.